src/levanter/compat/torch_serialization.py

```diff
@@ -366,22 +366,46 @@ def to_numpy_state_dict(model, prefix: Optional[str] = None) -> StateDict:
     This method is especially useful for saving models distributed across multiple hosts.
     """
 
-    def get_to_cpu(arr):
-        if not is_jax_array_like(arr):
-            return arr
-        elif isinstance(arr, np.ndarray):
-            return arr
-        elif arr.is_fully_addressable:
-            r = np.array(arr)
-            return r
-        else:
-            return np.array(jax.device_get(multihost_utils.process_allgather(arr, tiled=True)))
+    with jax.default_device(jax.local_devices(backend="cpu")[0]):
 
-    # need to make sure the model is on *this machine* and *this machine's CPU* before saving
-    model = jax.tree_map(lambda arr: get_to_cpu(arr), model)
-    # TODO: it's be nice if safetensors supported an iterator or something so we could do the allgather one at a time
-    state_dict = model.to_state_dict(prefix=prefix)
-    return state_dict
+        def get_to_cpu(arr):
+            if not is_jax_array_like(arr):
+                return arr
+            elif isinstance(arr, np.ndarray):
+                return arr
+            elif arr.is_fully_addressable:
+                r = np.array(arr)
+                return r
+            else:
+                # unfortunately, jax's allgather seems to replicate to every device rather than every host
```

Comment:
What's the relationship between device and host? 

From [TPU's system architecture](https://cloud.google.com/tpu/docs/system-architecture-tpu-vm#tpu_device), I found that:
> A TPU device contains four TPU chips interconnected by high-speed interconnects. 
> ...
> Each TPU device is connected to a CPU-based host machine (TPU VM). 

Does it mean 1 host contains 1 device, or not necessary?

Reply:
8 devices per host (on tpu v3)